### PR TITLE
Fix initialization of LA::d::Vector in MGTransfer::copy_to_mg

### PR DESCRIPTION
**Pull request by kronbichler**
As described in the change log, there could be a bug saying that we had ghost entries in a `LA::distributed::Vector` in a `MatrixFree::loop` where we called `Vector::compress()` and were not supposed to have ghosts. This happened due to a chain of decisions that all went the wrong direction. The code is designed to tolerate quite a bit, so that it went wrong was simply bad luck in the end. The code failed in the second matrix-vector product in the conjugate gradient method for estimating the eigenvalues for Chebyshev. What happened was:
- `DiagonalMatrix` used as an inner preconditioner in Chebyshev calls `dst = src;`, where `src` is a properly initialized matrix-free compatible vector, but `dst` is initialized only from the locally owned elements without ghost. In that scenario, the copy operation automatically makes `src` to update the ghost values, see here: https://github.com/dealii/dealii/blob/master/include/deal.II/lac/la_parallel_vector.templates.h#L670-L687
- The vector without ghosts comes from the initialization of the `MGLevelGlobalTransfer` class, which uses a separate function where there is no need to. This separate path dates back to before we had a template specialization. So we can delete some code to fix this problem.
- The `MatrixFree::cell_loop` was supposed to zero the vector, rather than doing it by `Vector::operator=()`.

### CODE REVIEW

File: tests/matrix_free/parallel_multigrid_adaptive_08.cc [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+      do_test<dim, fe_degree, fe_degree + 1, double>(dof);
+    }
+}
+
+
+
+int
+main(int argc, char **argv)
+{
+  Utilities::MPI::MPI_InitFinalize mpi_init(argc, argv, 1);
```
Comment (tjhei) [inline tests/matrix_free/parallel_multigrid_adaptive_08.cc]: can we use  MPILogInitAll?

**Comment by kronbichler** [comment]
I have to look into some other matrix-free multigrid tests - this patch seems to contradict some of them. I need to investigate this tomorrow.

**Comment by kronbichler** [comment]
It seems I only forgot to move the check whether we have the right local size to the new code position. `ctest -R matrix_free` works, I run the full suite over night.

**Comment by masterleinad** [comment]
`4228 - multigrid/transfer_06.mpirun=3.debug (Failed)` Is still failing on the serial tester.

**Comment by kronbichler** [comment]
Sorry for the noise. Apparently there was a reason behind some of the old indirections - we do not fill the ghosted level vectors for `MGTransferPrebuilt`, only for `MGTransferMatrixFree`. I have fixed it now. Now I essentially move the old code from a separate templated function to the position where we use it, plus add an indirection to capture the case considered in the test.

**Comment by kronbichler** [comment]
> is there a chance for this generating silently wrong results or would it always crash in debug mode?

In debug mode it would always crash, but in release mode it would have given wrong results (and be somewhat slower than it needed to, since it updated ghost values in places where there shouldn't be any).

**Pull request merged**